WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Categories/ProductCategoryViewModelBuilderTests.swift

```diff
@@ -0,0 +1,120 @@
+@testable import Yosemite
+
+/// Tests for `ProductCategoryViewModelBuilderTests`.
+///
+final class ProductCategoryViewModelBuilderTests: XCTestCase {
+
+    /// Sample siteID
+    ///
+    private let sampleSiteID: Int64 = 123
+
+    func testCategoriesWithoutSubCategoriesPreserveInitialOrder() {
+        // Given
+        let categories = sampleCategories(initialID: 1, count: 10)
+
+        // When
+        let viewModels = ProductCategoryViewModelBuilder.viewModels(from: categories, selectedCategories: [])
+
+        // Then
+        let expectedCategoriesNames = categories.map { $0.name }
+        let viewModelsNames = viewModels.map { $0.name }
+        XCTAssertEqual(expectedCategoriesNames, viewModelsNames)
+    }
+
+    func testCategoriesWithSubCategoriesAreFlattenedCorrectly() {
+        // Given
+        let rootCategories = sampleCategories(initialID: 1, count: 10)
+        let subCategories1 = sampleCategories(initialID: 11, parentID: 5, count: 5)
+        let subCategories2 = sampleCategories(initialID: 16, parentID: 13, count: 3)
+        let allCategories = rootCategories + subCategories1 + subCategories2
+
+        // When
+        let viewModels = ProductCategoryViewModelBuilder.viewModels(from: allCategories, selectedCategories: [])
+
+        // Then
+        let expectedCategoriesNames: [String] = {
+            var expected = rootCategories
+            expected.insert(contentsOf: subCategories1, at: 5)
+            expected.insert(contentsOf: subCategories2, at: 8)
+            return expected.map { $0.name }
+        }()
+        let viewModelsNames = viewModels.map { $0.name }
+        XCTAssertEqual(expectedCategoriesNames, viewModelsNames)
+    }
+
+    func testFlattenedCategoriesHaveACorrectIndentationLevel() {
+        // Given
+        let rootCategories = sampleCategories(initialID: 1, count: 10)
+        let subCategories1 = sampleCategories(initialID: 11, parentID: 5, count: 5)
+        let subCategories2 = sampleCategories(initialID: 16, parentID: 13, count: 3)
+        let allCategories = rootCategories + subCategories1 + subCategories2
+
+        // When
+        let viewModels = ProductCategoryViewModelBuilder.viewModels(from: allCategories, selectedCategories: [])
+
+        // Then
+        let rootCategoriesNames = rootCategories.map { $0.name }
+        let subCategories1Names = subCategories1.map { $0.name }
+        let subCategories2Names = subCategories2.map { $0.name }
+        for viewModel in viewModels {
+            if rootCategoriesNames.contains(viewModel.name) {
+                XCTAssertEqual(viewModel.indentationLevel, 0)
+            }
+
+            if subCategories1Names.contains(viewModel.name) {
+                XCTAssertEqual(viewModel.indentationLevel, 1)
+            }
+
+            if subCategories2Names.contains(viewModel.name) {
+                XCTAssertEqual(viewModel.indentationLevel, 2)
+            }
+        }
+    }
+
+    func testRootCategoriesAreMarkedAsSelected() {
```

Comment:
This name might be misleading because we're not really selecting the root categories right away, right?

Reply:
Do you mean something like `testRootCategoriesAreSelectedWhenProvidingProductSelectedCategories()`?

Reply:
I'd say that's better. This is challenging to name because, in the end, it's the same as `testSubCategoriesAreMarkedAsSelectec` below.